wake/testing/fuzzing/fuzz_shrink.py

```diff
@@ -0,0 +1,759 @@
+    test_instance = test_class()
+
+    flows: List[Callable] = __get_methods(test_instance, "flow")
+    invariants: List[Callable] = __get_methods(test_instance, "invariant")
+    shrank_path = get_shrank_path()
+    if shrank_path is None:
+        raise Exception("Shrunken data file path not found")
+    with open(shrank_path, 'rb') as f:
+            store_data: ShrankInfoFile = pickle.load(f)
+
+    random.setstate(pickle.loads(store_data.initial_state))
+    test_instance._flow_num = 0
+    test_instance._sequence_num = 0
+    test_instance.pre_sequence()
+
+    invariant_periods: DefaultDict[Callable[[None], None], int] = defaultdict(
+        int
+    )
+    for j in range(len(store_data.required_flows)):
+
+        flow = next((flow for flow in flows if store_data.required_flows[j].flow_name == flow.__name__), None)
+        if flow is None:
+            raise Exception("Flow not found")
+        flow_params = store_data.required_flows[j].flow_params
+        test_instance._flow_num = store_data.required_flows[j].flow_num
+        if not hasattr(flow, "precondition") or getattr(flow, "precondition")(test_instance):
+            random.setstate(pickle.loads(store_data.required_flows[j].random_state))
+            test_instance.pre_flow(flow)
+            flow(test_instance, *flow_params)
+            test_instance.post_flow(flow)
+
+        try:
+            random.setstate(pickle.loads(store_data.required_flows[j].before_inv_random_state))
+        except Exception as e:
+            pass
+
+        if not dry_run:
+            test_instance.pre_invariants()
+            for inv in invariants:
+                if invariant_periods[inv] == 0:
+                    test_instance.pre_invariant(inv)
+                    inv(test_instance)
+                    test_instance.post_invariant(inv)
+                invariant_periods[inv] += 1
+                if invariant_periods[inv] == getattr(inv, "period"):
+                    invariant_periods[inv] = 0
+            test_instance.post_invariants()
+
+    print("Shrunken test passed")
+
+def shrink_collecting_phase(test_instance: FuzzTest, flows, invariants, flow_states:List[FlowState], chains: Tuple[Chain, ...], flows_count: int) -> Tuple[Exception, timedelta]:
+    data_time = datetime.now()
+    flows_counter: DefaultDict[Callable, int] = defaultdict(int)
+    invariant_periods: DefaultDict[Callable[[None], None], int] = defaultdict(int)
+    # Snapshot all connected chains
+    initial_chain_state_snapshots = [chain.snapshot() for chain in chains]
+    error_flow_num = get_error_flow_num() # argument
+    initial_state = get_sequence_initial_internal_state() # argument
+
+    random.setstate(pickle.loads(initial_state))
+    with print_ignore():
+        test_instance._flow_num = 0
+        test_instance.pre_sequence()
+        exception_content = None
+        try:
+            with redirect_stdout(open(os.devnull, 'w')), redirect_stderr(open(os.devnull, 'w')):
```

Comment:
isn't the redirection applied twice? once in `print_ignore` and for the second time directly here?

Reply:
This was redundant. removed.